src/js/node/events.js

```diff
@@ -354,9 +357,55 @@ function once(emitter, type, options) {
   });
 }
 
-function on(emitter, type, options) {
-  var { signal, close, highWatermark = Number.MAX_SAFE_INTEGER, lowWatermark = 1 } = options || {};
-  throwNotImplemented("events.on", 2679);
+async function* on(emitter, event, options = {}) {
+  const signal = options.signal;
+  if (signal?.aborted) throw new AbortError(undefined, { cause: signal?.reason });
+
+  const unconsumedPromises = new FixedQueue();
+  const unconsumedEvents = new FixedQueue();
+  const unconsumedErrors = new FixedQueue();
+  let done = false;
+
+  emitter.on(event, ev => {
+    if (!unconsumedPromises.isEmpty()) {
+      const { resolve } = unconsumedPromises.shift();
+      return resolve([ev]);
+    }
+    unconsumedEvents.push([ev]);
+  });
+  emitter.on("error", ex => {
+    if (!unconsumedPromises.isEmpty()) {
+      const { reject } = unconsumedPromises.shift();
+      return reject(ex);
+    }
+    unconsumedErrors.push(ex);
+  });
+  signal?.addEventListener("abort", () => {
+    emitter.emit("error", new AbortError(undefined, { cause: signal?.reason }));
+  });
+
+  for (const evName of options?.close || []) {
+    emitter.on(evName, () => {
+      emitter.emit(event, undefined);
+      done = true;
+    });
+  }
+
+  while (!done) {
+    if (!unconsumedEvents.isEmpty()) {
+      yield new Promise(function (resolve, reject) {
```

Comment:
This code can be simplified.

```suggestion
      yield Promise.$resolve(unconsumedEvents.shift());
```